README
NAME
    Gentoo::Util::MetaCPAN - Gentoo Specific MetaCPAN Utilities.

VERSION
    version 0.001001

METHODS
  find_files_providing
  find_latest_files_providing
  find_release
  find_releases_providing
FUNCTIONS
  "mcpan"
AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Gentoo-Util-MetaCPAN

0.001001 2014-10-31T20:14:17Z 8389a03

0.001000 2014-10-31T20:09:38Z abf9e25
 - First version.



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.024.
use strict;
use warnings;

use 5.013002;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Gentoo Specific MetaCPAN Utilities.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Gentoo-Util-MetaCPAN",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.013002",
  "NAME" => "Gentoo::Util::MetaCPAN",
  "PREREQ_PM" => {
    "CHI::Config" => 0,
    "CPAN::Meta::Prereqs" => 0,
    "File::Spec" => 0,
    "Gentoo::PerlMod::Version" => 0,
    "HTTP::Tiny::Mech" => 0,
    "LWP::UserAgent" => 0,
    "MetaCPAN::Client" => 0,
    "MetaCPAN::Client::File" => 0,
    "MetaCPAN::Client::Release" => "1.007001",
    "MetaCPAN::Client::ResultSet" => 0,
    "Moo" => 0,
    "MooX::Lsub" => 0,
    "Path::Tiny" => 0,
    "Scalar::Util" => 0,
    "Sub::Exporter::Progressive" => 0,
    "WWW::Mechanize::Cached" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "List::UtilsBy" => 0,
    "Test::More" => "0.89",
    "Test::RequiresInternet" => 0
  },
  "VERSION" => "0.001001",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "CHI::Config" => 0,
  "CPAN::Meta::Prereqs" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Gentoo::PerlMod::Version" => 0,
  "HTTP::Tiny::Mech" => 0,
  "LWP::UserAgent" => 0,
  "List::UtilsBy" => 0,
  "MetaCPAN::Client" => 0,
  "MetaCPAN::Client::File" => 0,
  "MetaCPAN::Client::Release" => "1.007001",
  "MetaCPAN::Client::ResultSet" => 0,
  "Moo" => 0,
  "MooX::Lsub" => 0,
  "Path::Tiny" => 0,
  "Scalar::Util" => 0,
  "Sub::Exporter::Progressive" => 0,
  "Test::More" => "0.89",
  "Test::RequiresInternet" => 0,
  "WWW::Mechanize::Cached" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw( MetaCPAN ));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
$bundle->remove_policy('CodeLayout::RequireUseUTF8');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/Gentoo/Util/MetaCPAN/File.pm
use 5.006;
use strict;
use warnings;

package Gentoo::Util::MetaCPAN::File;

our $VERSION = '0.001001'; # TRIAL

# ABSTRACT: Enhancements to MetaCPAN::File

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( extends has around );
use Scalar::Util qw( blessed );
extends 'MetaCPAN::Client::File';

has latest => ( is => ro =>, lazy => 1, default => sub { $_[0]->data->{latest} } );

around _known_fields => sub {
  my ( $orig, $self, @args ) = @_;
  return [ ( 'latest', @{ $self->$orig(@args) } ) ];
};

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Gentoo::Util::MetaCPAN::File - Enhancements to MetaCPAN::File

=head1 VERSION

version 0.001001

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Gentoo/Util/MetaCPAN/Release.pm
use 5.006;
use strict;
use warnings;

package Gentoo::Util::MetaCPAN::Release;

our $VERSION = '0.001001'; # TRIAL

# ABSTRACT: Subclass of MetaCPAN::Client::Release with some utility functions

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( extends has around );
use Gentoo::PerlMod::Version qw( gentooize_version );
use Gentoo::Util::MetaCPAN::Requirement;
use CPAN::Meta::Prereqs;
use MetaCPAN::Client::Release 1.007001;
use MetaCPAN::Client::ResultSet;

extends 'MetaCPAN::Client::Release';

has 'prereqs' => ( is => ro =>, lazy => 1, builder => '_build_prereqs' );

sub _build_prereqs {
  my ($self) = @_;
  if ( exists $self->metadata->{prereqs} ) {
    return CPAN::Meta::Prereqs->new( $self->metadata->{prereqs} );
  }
  my $pre = CPAN::Meta::Prereqs->new();

  for my $dependency ( @{ $self->{data}->{dependency} } ) {
    my $stash = $pre->requirements_for( $dependency->{phase}, $dependency->{relationship} );
    $stash->add_string_requirement( $dependency->{module}, $dependency->{version} );
  }
  return $pre;
}

sub get_dependencies {
  my ( $self, $phases, $relationships ) = @_;
  my $req = $self->prereqs->merged_requirements( $phases, $relationships );
  my @out;
  for my $module ( sort $req->required_modules ) {
    push @out,
      Gentoo::Util::MetaCPAN::Requirement->new(
      module => $module,
      range  => $req->{requirements}->{$module},
      );
  }
  return \@out;
}

sub gentoo_version {
  my ($self) = @_;
  ## no critic (ErrorHandling::RequireCheckingReturnValueOfEval)
  return eval { gentooize_version( $self->version, { lax => 1 } ) };
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Gentoo::Util::MetaCPAN::Release - Subclass of MetaCPAN::Client::Release with some utility functions

=head1 VERSION

version 0.001001

=head1 METHODS

=head2 gentoo_version

=head2 get_dependencies

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Gentoo/Util/MetaCPAN/Requirement.pm
use 5.006;
use strict;
use warnings;

package Gentoo::Util::MetaCPAN::Requirement;

our $VERSION = '0.001001'; # TRIAL

# ABSTRACT: A Single dependency requirement specialized for Gentoo

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( has );
use Gentoo::PerlMod::Version qw( gentooize_version );

has 'module'         => ( is => ro =>, required => 1, );
has 'range'          => ( is => ro =>, required => 1, );
has 'version'        => ( is => ro =>, lazy     => 1, builder => '_build_version' );
has 'gentoo_version' => ( is => ro =>, lazy     => 1, builder => '_build_gentoo_version' );

sub BUILD {
  my ($self) = @_;
  $self->gentoo_version;
  return;
}

sub _has_min_version {
  my ($self) = @_;
  ## no critic (Subroutines::ProtectPrivateSubs)
  return !( $self->range->_accepts(0) );
}

sub _build_version {
  my ($self) = @_;
  return $self->range->as_string;
}

sub _build_gentoo_version {
  my ($self) = @_;
  my $ver;
  ## no critic (ErrorHandling::RequireCheckingReturnValueOfEval)
  eval { $ver = gentooize_version( $self->version, { lax => 1 } ); };
  return $ver;
}

sub _pretty_version {
  my ($version) = @_;
  return eval { gentooize_version( $version, { lax => 1 } ) } || q<?>;
}

sub _pretty {
  my ($self) = @_;
  my $module = $self->module;
  my $range  = $self->range;
  if ( $range->isa('CPAN::Meta::Requirements::_Range::Exact') ) {
    return sprintf '= %s @ %s', $module, _pretty_version( $range->{version} );
  }
  my @out;
  if ( exists $range->{minimum} ) {
    if ( $range->{minimum} != 0 ) {
      push @out, sprintf '>= %s @ %s', $module, _pretty_version( $range->{minimum} );
    }
    else {
      push @out, sprintf '%s', $module;
    }
  }
  if ( exists $range->{maximum} ) {
    push @out, sprintf '<= %s @ %s', $module, _pretty_version( $range->{maximum} );
  }
  if ( exists $range->{exclusions} ) {
    for my $exclusion ( @{ $range->{exclusions} } ) {
      push @out, sprintf '! %s @ %s', $module, _pretty_version( $exclusion, { lax => 1 } );
    }
  }
  return join q[ ], @out;
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Gentoo::Util::MetaCPAN::Requirement - A Single dependency requirement specialized for Gentoo

=head1 VERSION

version 0.001001

=head1 METHODS

=head2 BUILD

=head2 module

=head2 range

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Gentoo/Util/MetaCPAN.pm
use 5.006;
use strict;
use warnings;

package Gentoo::Util::MetaCPAN;

our $VERSION = '0.001001'; # TRIAL

# ABSTRACT: Gentoo Specific MetaCPAN Utilities.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo;
use MooX::Lsub qw( lsub );
use File::Spec;
use Sub::Exporter::Progressive -setup => { exports => ['mcpan'] };
use Path::Tiny qw( path );
use CHI::Config;

lsub '_caches' => sub {
  return CHI::Config->new(
    defaults => [
      {
        type   => "driver",
        name   => "gentoo-util-metacpan.web",
        config => {
          driver    => "Memory",
          datastore => {},
        },
      },
      {
        type   => "driver",
        name   => "gentoo-util-metacpan.objects",
        config => {
          driver    => "Memory",
          datastore => {},
        },
      },
    ]
  );
};

lsub 'www_cache'    => sub { $_[0]->_caches->get_cache('gentoo-util-metacpan.web') };
lsub 'object_cache' => sub { $_[0]->_caches->get_cache('gentoo-util-metacpan.objects') };

lsub 'debug' => sub {
  return unless defined $ENV{WWW_MECH_DEBUG};
  return $ENV{WWW_MECH_DEBUG};
};
lsub 'nocache' => sub {
  return unless defined $ENV{WWW_MECH_NOCACHE};
  return $ENV{WWW_MECH_NOCACHE};
};
lsub 'mechua' => sub {
  my ($self) = @_;
  my $mech;
  if ( $self->nocache ) {
    require LWP::UserAgent;
    $mech = LWP::UserAgent->new();
  }
  else {
    require WWW::Mechanize::Cached;
    $mech = WWW::Mechanize::Cached->new(
      cache     => $self->www_cache,
      timeout   => 20_000,
      autocheck => 1,
    );
  }
  if ( ( $self->debug || 0 ) > 1 ) {
    $mech->add_handler(
      'request_send' => sub {
        *STDERR->printf( "%s\n", $_[0]->as_string );
        return;
      },
    );
    $mech->add_handler(
      'response_done' => sub {
        *STDERR->printf( "%s\n", $_[0]->content );
        return;
      },
    );
  }
  elsif ( $self->debug ) {
    $mech->add_handler(
      'request_send' => sub {
        *STDERR->printf( "%s\n", $_[0]->dump );
        return;
      },
    );
    $mech->add_handler(
      'response_done' => sub {
        *STDERR->printf( "%s\n", $_[0]->dump );
        return;
      },
    );
  }
  return $mech;
};
lsub 'tinymech' => sub {
  my ($self) = @_;
  require HTTP::Tiny::Mech;
  HTTP::Tiny::Mech->new( mechua => $self->mechua );
};
lsub 'client' => sub {
  my ($self) = @_;
  require MetaCPAN::Client;
  MetaCPAN::Client->new( ua => $self->tinymech );
};

sub _cache_object {
  my ( $self, $key, $time, $code ) = @_;
  if ( $self->nocache ) {
    return $code->();
  }
  return $self->object_cache->compute( $key, $time, $code );
}
{
## HACK: This exists because its not supported natively yet.
  ## no critic (TestingAndDebugging::ProhibitNoWarnings,Subroutines::ProhibitQualifiedSubDeclarations)
  no warnings 'redefine';

  use MetaCPAN::Client::ResultSet;

  sub MetaCPAN::Client::ResultSet::next {
    my $self = shift;
    my $result =
        $self->has_scroller
      ? $self->scroller->next
      : shift @{ $self->items };

    defined $result or return;

    my $class = exists $self->{'class'} ? $self->{class} : 'MetaCPAN::Client::' . ucfirst $self->type;
    return $class->new_from_request( $result->{'_source'} || $result->{'fields'} );
  }
}

# More hacks because the native MetaCPAN client is a bit broken
sub _raw_scroll_query {
  my ( $self, $config ) = @_;

  my $creq = $self->client->request;

  my $class = delete $config->{'class'};
  my $type  = delete $config->{type};

  my $scroller = $creq->ssearch( $type, { bogus => 1 }, $config );

  if ( not $class ) {
    $class = 'MetaCPAN::Client::' . ucfirst( $config->{type} );
  }
  my $rs = MetaCPAN::Client::ResultSet->new(
    type     => $type,
    scroller => $scroller,
  );
  $rs->{class} = $class;
  return $rs;
}

sub _scroll_to_list {
  my ( undef, $scroll ) = @_;
  my @out;
  while ( my $item = $scroll->next ) {
    push @out, $item;
  }
  return @out;
}

sub find_release {
  my ( $self, $author, $dist ) = @_;
  require Gentoo::Util::MetaCPAN::Release;
  my $query = {
    type  => 'release',
    class => 'Gentoo::Util::MetaCPAN::Release',
    body  => {
      query => {
        bool => {
          must => [
            { term => { name   => $dist } },      #
            { term => { author => $author } },    #
          ],
        },
      },
    },
  };
  my $result = $self->_cache_object(
    [ 'find_release', $author, $dist ] => undef,
    => sub {
      return [ $self->_scroll_to_list( $self->_raw_scroll_query($query) ), ];
    },
  );
  return @{$result};
}

sub find_files_providing {
  my ( $self, $module_name ) = @_;
  require Gentoo::Util::MetaCPAN::File;

  my @terms;
  push @terms, { term => { 'module.authorized' => 1 } };
  push @terms, { term => { 'module.indexed'    => 1 } };
  push @terms, { term => { 'module.name'       => $module_name } };

  my $nested_query = {
    constant_score => {
      filter => {
        and => \@terms,
      },
    },
  };
  my $query = {
    filtered => {
      query => {
        nested => {
          path  => 'module',
          query => $nested_query,
        },
      },
    },
  };
  my $config = {
    type  => 'file',
    class => 'Gentoo::Util::MetaCPAN::File',
    body  => {
      query => $query,
    },
  };
  my $result = $self->_cache_object(
    [ 'find_files_providing', $module_name ] => undef,
    => sub {
      return [ $self->_scroll_to_list( $self->_raw_scroll_query($config) ) ];
    },
  );
  return @{$result};
}

sub find_latest_files_providing {
  my ( $self, $module_name ) = @_;
  require Gentoo::Util::MetaCPAN::File;

  my @terms;
  push @terms, { term => { 'module.authorized' => 1 } };
  push @terms, { term => { 'module.indexed'    => 1 } };
  push @terms, { term => { 'module.name'       => $module_name } };

  my $nested_query = {
    constant_score => {
      filter => {
        and => \@terms,
      },
    },
  };
  my $query = {
    filtered => {
      query => {
        nested => {
          path  => 'module',
          query => $nested_query,
        },
      },
    },
  };
  my $config = {
    type  => 'file',
    class => 'Gentoo::Util::MetaCPAN::File',

    body => {
      fields        => q[*],
      script_fields => { latest => { 'metacpan_script' => 'status_is_latest' } },
      query         => $query,
    },
  };
  my $result = $self->_cache_object(
    [ 'find_latest_files_providing', $module_name ] => undef,
    => sub {
      return [ grep { $_->latest } $self->_scroll_to_list( $self->_raw_scroll_query($config) ) ];
    },
  );
  return @{$result};

}

sub find_releases_providing {
  my ( $self, $module_name ) = @_;
  require Gentoo::Util::MetaCPAN::Release;

  my $nested_query = {
    bool => {
      must => [
        { term => { 'authorized' => 1 } },               #
        { term => { 'indexed'    => 1 } },               #
        { term => { 'name'       => $module_name } },    #
      ],
    },
  };

  my $query = {
    nested => {
      path  => 'module',
      query => $nested_query,
    },
  };

  my $config = {
    type  => 'release',
    class => 'Gentoo::Util::MetaCPAN::Release',
    body  => {
      query => {    #     %{$query},
        constant_score => { query => $query },
      },
    },
  };
  my $result = $self->_cache_object(
    [ 'find_releases_providing', $module_name ] => undef,
    => sub {
      return [ $self->_scroll_to_list( $self->_raw_scroll_query($config) ) ];
    },
  );
  return @{$result};
}

sub mcpan { return __PACKAGE__->new() }

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Gentoo::Util::MetaCPAN - Gentoo Specific MetaCPAN Utilities.

=head1 VERSION

version 0.001001

=head1 METHODS

=head2 find_files_providing

=head2 find_latest_files_providing

=head2 find_release

=head2 find_releases_providing

=head1 FUNCTIONS

=head2 C<mcpan>

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
